### PR TITLE
Container Registry auth token URI auto-retrieval

### LINKED ISSUE #4178: Improve the Container Registry authorization token URI auto-retrieval
**Problem description:**
For stream and task applications loaded from private Container Registries, we need to provide DataFlow with  [RegistryConfiguration](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/RegistryConfiguration.java) configuration to allow the server pull the metadata for those applications.
 
In case of the OAuth2 authorization the registry configuration look like this:
```
- spring.cloud.dataflow.container.registry-configurations[harbor].registry-host=demo.goharbor.io
- spring.cloud.dataflow.container.registry-configurations[harbor].authorization-type=dockeroauth2
- spring.cloud.dataflow.container.registry-configurations[harbor].user=***
- spring.cloud.dataflow.container.registry-configurations[harbor].secret=***
- spring.cloud.dataflow.container.registry-configurations[harbor].extra[registryAuthUri]=https://demo.goharbor.io/service/token?service=harbor-registry&scope=repository:{repository}:pull
```

The `registryAuthUri` configuration parameter sets the authorization token URI endpoint used by the API to obtain a valid token used for interacting with the repository.  Instead of setting this parameter explicitly there is [getDockerTokenServiceUri](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-configuration-metadata/src/main/java/org/springframework/cloud/dataflow/configuration/metadata/container/authorization/DockerConfigJsonSecretToRegistryConfigurationConverter.java#L106) method that can compute this property form the HTTP 401 response headers. 

Current `getDockerTokenServiceUri` implementation probes the `https://<registry-host>/v2/_catalog` to obtain HTTP 401 response.  For some registries this url could be less reliable to obtain OAuth2 Error HTTP response. Instead the `https://<registry-host>/v2/` should be used.

### PR DESCRIPTION
**Pull request by tzolov**
- Replace the probe url from https://<registry-host/v2/_catalog to https://<registry-host/v2/
  - Remove the base auth when hitting the probe registry url.
  - Improve the error handling. Check of potential NPE and throw AppMetadataResolutionException instead.
  - Reduce the logging messages in case of registry configuration problem. Don't print the exception traces for non debug level.
  - Minor code improvements.

  Resolves #4178

### CODE REVIEW

File: src/kubernetes/server/server-deployment.yaml [hunk 1]
```diff
@@ -64,7 +67,7 @@ spec:
         - name: SPRING_CLOUD_KUBERNETES_SECRETS_ENABLE_API
           value: 'false'
         - name: SPRING_CLOUD_KUBERNETES_SECRETS_PATHS
-          value: /etc/secrets
+          value: /etc/secrets,/workspace/runtime/secrets/ctr-registry
```
Comment (chrisjs) [inline src/kubernetes/server/server-deployment.yaml]: probably not needed with the above change

**Review by chrisjs — CHANGES_REQUESTED** [review]

File: src/kubernetes/server/server-deployment.yaml [hunk 2]
```diff
@@ -25,6 +25,9 @@ spec:
           - name: database
             mountPath: /etc/secrets/database
             readOnly: true
+          - name: ctr-registry
+            mountPath: /workspace/runtime/secrets/ctr-registry
```
[the remark this answers is not in the document]
Comment (tzolov) [reply src/kubernetes/server/server-deployment.yaml]: This was mistakenly committed. I removed it form the PR. The  server-deployment.yaml should be changed by this PR

File: src/kubernetes/server/server-deployment.yaml (hunk 1 above)
Comment (tzolov) [reply src/kubernetes/server/server-deployment.yaml]: same, this was mistakenly committed and I removed it form the PR. The  server-deployment.yaml should be changed by this PR

**Comment by chrisjs** [comment]
merged in 2f7ef64cacd075d6faf47dd5ee3cbe497b3dfeb4 & backported in 168cb9ebb81b1242894977a68707c133f35e65d9

**Pull request closed without merging**